# **Waze**

# **Inspect and analyze data**

**The purpose** of this project is to investigate and understand the Waze data set provided.

*This Notebook has three parts:*

**Part 1:** Understand the situation
* How can one best prepare to understand and organize the provided information?

**Part 2:** Understand the data

* Create a pandas dataframe for data learning, future exploratory data analysis (EDA), and statistical activities

* Compile summary information about the data to inform next steps

**Part 3:** Understand the variables

* Use insights from the examination of the summary data to guide deeper investigation into variables


<br/>


# **PACE stages**

## **PACE: Plan**

### **Task 1. Understand the situation**


Steps to follow:

* Explore the relations between the different columns of the dataframes.
* Analyze the outliners and explore the means and medians.
* See if there are non random causes that can affect the distribution of missing values.
* Compare percentages of the general dataset and those around specific values, such as % of churn in the general view vs. % of churn among the non null values.
* Think outside the box: check if the data from the drivers correspond to a normal user or to super drivers.


## **PACE: Analyze**

### **Task 2a. Imports and data loading**

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
file_path = '/content/drive/MyDrive/Data Analytics/Main Projects/Waze project/waze_dataset.csv'
df = pd.read_csv(file_path)

### **Task 2b. Summary information**

In [ ]:
df.head()

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


In [ ]:
df["label"].value_counts()

retained    11763
churned      2536
Name: label, dtype: int64

In [ ]:
label_null = df[df["label"].isnull()]
label_null

ID label  sessions  drives  total_sessions  n_days_after_onboarding  \
77        77   NaN        63      50      133.104155                      783   
80        80   NaN       116      93      436.060183                     1584   
98        98   NaN        78      64      583.492789                     3414   
111      111   NaN       106     102      113.379056                     2228   
142      142   NaN        32      26      222.129310                      208   
...      ...   ...       ...     ...             ...                      ...   
14941  14941   NaN       191     160      485.328204                     1287   
14943  14943   NaN        48      38       96.797017                      555   
14945  14945   NaN        34      29      134.416604                     1643   
14972  14972   NaN       220     181      256.212166                     1718   
14993  14993   NaN        67      57       97.570074                     1131   

       total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
77                        201                       0       2649.015822   
80                        283                      62       4183.409514   
98                          0                       0       1811.140893   
111                        14                       0       2817.481840   
142                        55                      10       2459.816477   
...                       ...                     ...               ...   
14941                      25                       0       6468.181924   
14943                       0                       6       8266.129497   
14945                     268                       2       4554.007843   
14972                     360                      23       5586.913459   
14993                     207                     102       2267.052913   

       duration_minutes_drives  activity_days  driving_days   device  
77                 1517.209970             19            13   iPhone  
80                 3121.889952             18            15   iPhone  
98                  642.189122             12            11  Android  
111                2011.724274             17            13  Android  
142                 874.427617             11             7   iPhone  
...                        ...            ...           ...      ...  
14941              3466.104564             14            14   iPhone  
14943              5902.351711             19            19   iPhone  
14945              1579.211201             18            17  Android  
14972              4104.440202             19            18   iPhone  
14993               318.120634             27            26   iPhone  

[700 rows x 13 columns]

### **Task 2c. Null values and summary statistics**

Comparing the summary statistics of the 700 rows that are missing labels with summary statistics of the rows that are not missing any values:

In [ ]:
# Isolate rows with null values
print(label_null.shape)
label_null.describe()

(700, 13)


ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [ ]:
# Isolate rows without null values
non_null = df[df["label"].notnull()]
print(non_null.shape)
non_null.describe()

(14299, 13)


ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000


**Question:** Is there a discernible difference between the two populations?

1. There are no visible differences between the dataframes. The mean, median and std are similiar among the two populations.

### **Task 2d. Null values - device counts**

Checking the two populations with respect to the `device` variable:

In [ ]:
# Count of null values by device
null_devices = label_null[["device"]]
null_devices.value_counts()

device 
iPhone     447
Android    253
dtype: int64

**Question:** How many iPhone users had null values and how many Android users had null values?

1. There are 447 iPhone users and 253 Android users with null values.

Calculating the percentage with each device, Android and iPhone:

In [ ]:
null_devices.value_counts(normalize=True)

device 
iPhone     0.638571
Android    0.361429
dtype: float64

How does this compare to the device ratio in the full dataset?

In [ ]:
df_devices = df[["device"]]
df_devices.value_counts(normalize=True)

device 
iPhone     0.644843
Android    0.355157
dtype: float64

* The percentage of missing values by each device is consistent with their representation in the data overall.

* There is nothing to suggest a non-random cause of the missing data.

###### The distribution of missing values among different devices resembles the general distribution of those devices in the dataset, and there are no clear indications of a non-random cause behind the missing data.

### **Task 2e. Non-null values analysis**

Examining the counts and percentages of users who churned vs. those who were retained:

How many of each group are represented in the data?

In [ ]:
non_null["label"].value_counts()

retained    11763
churned      2536
Name: label, dtype: int64

In [ ]:
non_null['label'].value_counts(normalize=True)

retained    0.822645
churned     0.177355
Name: label, dtype: float64

* This dataset contains 82% retained users and 18% churned users.

Next, comparing the medians of each variable for churned and retained users:
> The reason for calculating the median and not the mean is that one doesn't want outliers to unduly affect the portrayal of a typical user. Notice, for example, that the maximum value in the `driven_km_drives` column is 21,183 km. That's more than half the circumference of the earth.




In [ ]:
medians = non_null.groupby(["label"]).median(numeric_only = True)
medians

ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  
label                                                           
churned               1607.183785            8.0           6.0  
retained              1458.046141           17.0          14.0

This offers an interesting snapshot of the two groups, churned vs. retained:

Users who churned averaged ~3 more drives in the last month than retained users, but retained users used the app on over twice as many days as churned users in the same time period.

The median churned user drove ~200 more kilometers and 2.5 more hours during the last month than the median retained user.

It seems that churned users had more drives in fewer days, and their trips were farther and longer in duration. Perhaps this is suggestive of a user profile.

Calculating the median kilometers per drive in the last month for both retained and churned users:

In [ ]:
medians["km_per_drive_median"] = medians["driven_km_drives"]/medians["drives"]
medians[["km_per_drive_median"]]

km_per_drive_median
label                        
churned             73.053113
retained            73.716694

The median user from both groups drove ~73 km/drive. How many kilometers per driving day was this?

In [ ]:
medians["km_per_day"] = medians["driven_km_drives"]/medians["driving_days"]
medians[["km_per_day"]]

km_per_day
label               
churned   608.775944
retained  247.477472

Calculating the median number of drives per driving day for each group:

In [ ]:
# Divide the median number of drives by median number of driving days
medians["drives_per_driving day"] = medians["drives"]/medians["driving_days"]
medians[["drives_per_driving day"]]

drives_per_driving day
label                           
churned                 8.333333
retained                3.357143

The median user who churned drove 608 kilometers each day they drove last month, which is almost 250% the per-drive-day distance of retained users. The median churned user had a similarly disproporionate number of drives per drive day compared to retained users.

It is clear from these figures that, regardless of whether a user churned or not, the users represented in this data are serious drivers. It would probably be safe to assume that this data does not represent typical drivers at large. Perhaps the data&mdash;and in particular the sample of churned users&mdash;contains a high proportion of long-haul truckers.

In consideration of how much these users drive, it would be worthwhile to recommend to Waze that they gather more data on these super-drivers. It's possible that the reason for their driving so much is also the reason why the Waze app does not meet their specific set of needs, which may differ from the needs of a more typical driver, such as a commuter.

Examining whether there is an imbalance in how many users churned by device type:

In [ ]:
devices = non_null.groupby(["label","device"]).agg({"device":"count"})
devices

device
label    device         
churned  Android     891
         iPhone     1645
retained Android    4183
         iPhone     7580

In [ ]:
totals = devices.groupby(level=0).sum()
totals

device
label           
churned     2536
retained   11763

In [ ]:
percentages = devices.div(totals, level=0) * 100
percentages = percentages.rename(columns={'device': 'percentage'})
percentages

percentage
label    device             
churned  Android   35.134069
         iPhone    64.865931
retained Android   35.560656
         iPhone    64.439344

The ratio of iPhone users and Android users is consistent between the churned group and the retained group, and those ratios are both consistent with the ratio found in the overall dataset.

## **PACE: Construct**

**Note**: The Construct stage does not apply to this workflow. The PACE framework can be adapted to fit the specific requirements of any project.



## **PACE: Execute**



### **Task 3. Conclusion**

**Questions:**

1. Did the data contain any missing values? How many, and which variables were affected? Was there a pattern to the missing data?
    - Yes, there were 700 values missing, all inside the "label" variable. From theses, 447 are iPhone users and 253 Android users. There was not a pattern, the distribution of the missing data does not have clear indicator of a non-random cause.
    



2. What is a benefit of using the median value of a sample instead of the mean?
    - When we use the median, we can eliminate the outliners so they do not unduly affect the portrayal of a typical user.


3. Did your investigation give rise to further questions that you would like to explore or ask the Waze team about?
    - Yes, the investigation revealed that this dataset contains information about "superdrivers", users that drive realy long distances. It would be necessary to know them better to get to the cause for the churn.


4. What percentage of the users in the dataset were Android users and what percentage were iPhone users?
    - Androids users represent aprox. 36% and Android user 64%.


5. What were some distinguishing characteristics of users who churned vs. users who were retained?
    - Churned users have around 3 more drives, 200 more km driven, their drives long 2.5 hours more, they have half activity days, they drive 361 more km per day than retained users, and have 5 more drives per day.


6. Was there an appreciable difference in churn rate between iPhone users vs. Android users?
    - No, the percentages were similar.


